pkg/report/decompile.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package report
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/sys/targets"
+)
+
+type DecompilerFlag string
+
+// Extra flags that control the flow of decompilation.
+const (
+	FlagForceArmThumbMode DecompilerFlag = "force_arm_thumb"
+)
+
+const objdumpCallTimeout = 10 * time.Second
+
+type DecompiledOpcode struct {
+	Offset  int
+	Len     int
+	Command string
+	IsBad   bool
+	// Yes, the field below duplicates information and can be inferred from the other fields,
+	// but at this point it does not seem to be reasonable to parse a string and then build
+	// exactly that string from its parts.
+	FullDescription string
+}
+
+// Decompiles a byte array with opcodes into human-readable descriptions.
+// Target must specify the environment from which the opcodes were taken.
+func DecompileOpcodes(raw []byte, flags []DecompilerFlag, target *targets.Target) ([]DecompiledOpcode, error) {
+	args, err := objdumpBuildArgs(flags, target)
+	if err != nil {
+		return nil, err
+	}
+
+	outLines, err := objdumpExecutor(raw, args, target)
+	if err != nil {
+		return nil, err
+	}
+
+	list := objdumpParseOutput(outLines)
+	if len(list) == 0 && len(raw) > 0 {
+		return nil, fmt.Errorf("no instructions found while the total size is %v bytes", len(raw))
+	}
+	return list, nil
+}
+
+func objdumpExecutor(raw []byte, args []string, target *targets.Target) ([]string, error) {
+	file, err := ioutil.TempFile("", "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp file: %v", err)
+	}
+	defer os.Remove(file.Name())
+	defer file.Close()
+
+	_, err = file.Write(raw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write to temp file: %v", err)
+	}
+
+	args = append(args, file.Name())
+	cmd := osutil.Command(target.Objdump, args...)
+
+	respBytes, err := osutil.Run(objdumpCallTimeout, cmd)
```

Comment:
This function can be shorter:
```
	file, err := osutil.TempFile("syz-decompile")
	if err != nil {
		return nil, err
	}
	defer os.Remove()

	if err := osutil.WriteFile(file, raw); err != nil {
		return nil, fmt.Errorf("failed to write to temp file: %v", err)
	}

	return osutil.RunCmd(objdumpCallTimeout, "", target.Objdump, append(args, file)...)
```

Reply:
Is `os.Remove()` guaranteed to close the file handle on non-Unix OSes?

The example in the [official documentation](https://pkg.go.dev/io/ioutil#TempFile) still does Close().

Reply:
I used osutil.TempFile which returns file name instead of ioutil.TempFile.

Reply:
Ah, I didn't notice. Thanks for the clarification!